### PR TITLE
Improve reference cast help message

### LINKED ISSUE #37338: Help is not helpful for error
If you try to compile the code ([play link](https://play.rust-lang.org/?code=fn%20main%28%29%20%7B%0A%20%20%20%20vec!%5B0.0%5D.iter%28%29.map%28%7Cs%7C%20s%20as%20i16%29.collect%3A%3A%3CVec%3Ci16%3E%3E%28%29%3B%0A%7D&version=nightly&backtrace=0)):

``` rust
vec![0.0].iter().map(|s| s as i16).collect::<Vec<i16>>();
```

You get the error

```
rustc 1.14.0-nightly (6e8f92f11 2016-10-07)
error: casting `&f64` as `i16` is invalid
 --> <anon>:2:30
  |
2 |     vec![0.0].iter().map(|s| s as i16).collect();
  |                              ^^^^^^^^
  |
  = help: cast through a raw pointer first

error: aborting due to previous error
```

But casting through a raw pointer (`s as *const f64 as *const i16 as i16`) is not the idiomatic way IMO. Instead it should be fixed by simply dereferencing:

``` rust
vec![0.0].iter().map(|s| *s as i16).collect();
```

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #37338.

### CODE REVIEW

**Comment by eddyb** [comment]
You completely changed the meaning though, now it's bad if the user wanted to do the pointer thing.
cc @rust-lang/compiler

**Comment by GuillaumeGomez** [comment]
But at least it's not incorrect anymore.

**Comment by TimNN** [comment]
@GuillaumeGomez:

I'm not sure, but wouldn't your patch suggest dereferencing the pointer also in the following case:

```rust
let _ = &String::new() as usize;
```

Here, the original message is more accurate, I think.

**Comment by fhartwig** [comment]
In @TimNN's example, the original message clearly makes more sense. In the example from #37338, there is genuine ambiguity about which one the user wanted to do, but I would argue that the new suggestion is probably better because I suspect that the deref case is the more common one.

Would it makes sense to check the `RPtr`s type and only make the new suggestion if it's a numeric type? Or maybe suggest both solutions in that case, since there isn't a way to figure out which one the user meant.

**Comment by eddyb** [comment]
I'd argue that only `usize` is a resonable target for a pointer-to-integer cast, so we could just not suggest anything otherwise - simplest option - besides, @jonathandturner has some general type mismatch suggestion ideas which should apply here as well.

**Comment by GuillaumeGomez** [comment]
@TimNN: Well saw and not well saw at the same time. In this case, you should have to explicitly cast into a pointer before casting to usize. However, it just puts under the light a big issue: the compiler cannot guess what the user intends to do in some cases and only provide the simplest solution.

In this case, more people will certainly think that you just want to get the length and you put `&` by mistake.

@fhartwig: Still difficult. But we could provide both helps. What do you think about this @eddyb?

**Comment by eddyb** [comment]
@GuillaumeGomez IMO having both is more confusing.
Like I said, simplest thing to do is only provide the ptr help when `as usize` is being attempted.

**Comment by GuillaumeGomez** [comment]
Ok, I'll update it this way then.

**Comment by estebank** [comment]
@GuillaumeGomez, I hadn't seen that this PR was already out, so I took a stab at it as well. I still feel there's some value in PR #37442, as it handles the case @TimNN pushed forward as it is now, and specializes the cases of `&{integer}` and `&{float}` being cast to `{integer}` or `{float}`.

For a given file:

```rust
fn main() {
    vec![0.0].iter().map(|s| s as i16).collect::<Vec<i16>>();
    let _ = &String::new() as usize;
}
```
PR #37442 outputs:

```nocode
error: casting `&f64` as `i16` is invalid
 --> file3.rs:2:35
  |
2 |     vec![0.0].iter().map(|s| s as i16).collect::<Vec<i16>>();
  |                              -    ^^^ casting `&f64` as `i16` is invalid
  |                              |
  |                              try dereferencing for the cast to work

error: casting `&std::string::String` as `usize` is invalid
 --> file3.rs:3:13
  |
3 |     let _ = &String::new() as usize;
  |             ^^^^^^^^^^^^^^^^^^^^^^^ cast through a raw pointer first

error: aborting due to 2 previous errors
```

**Comment by GuillaumeGomez** [comment]
@estebank: I still need to update it anyway. I intend to do it this week-end.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by frewsxcv** [comment]
Travis failure looks unrelated.

```
Building stage2 tool linkchecker (x86_64-unknown-linux-gnu)
 Downloading url v1.2.2
 Downloading matches v0.1.3
 Downloading idna v0.1.0
 Downloading unicode-normalization v0.1.2
warning: spurious network error (2 tries remaining): [28] Timeout was reached
error: unable to get packages from source

Caused by:
  failed to verify the checksum of `unicode-normalization v0.1.2`

```

**Comment by GuillaumeGomez** [comment]
I restart it.

**Comment by GuillaumeGomez** [comment]
cc @eddyb

**Comment by eddyb** [comment]
You seem to have removed too many messages.

r? @jonathandturner

**Comment by jonathandturner** [comment]
Agree with @eddyb - why remove the help messages?

**Comment by GuillaumeGomez** [comment]
@jonathandturner: It comes after a discussion between @eddyb and me. Since a lot of messages were incorrect in some cases, it should be better to just remove them. However, after the last messages, I think I misunderstood him. Any further explanation would be appreciated. :)

**Comment by eddyb** [comment]
What I mean is only providing the "pointer to integer" help if `usize` is the target, i.e. hiding it for other integer types.

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/test/compile-fail/cast-rfc0401.rs [hunk 1]
```diff
@@ -54,7 +54,7 @@ fn main()
     //~^^ HELP through a usize first
     let _ = &v as usize;
     //~^ ERROR casting
-    //~^^ HELP through a raw pointer first
+    //~| HELP cast through a raw pointer first
```
Comment (eddyb) [inline src/test/compile-fail/cast-rfc0401.rs]: I don't think you need to change anything in this file.

File: src/test/compile-fail/fat-ptr-cast.rs [hunk 2]
```diff
@@ -18,11 +18,9 @@ fn main() {
     let q = a.as_ptr();
 
     a as usize; //~ ERROR casting
-    //~^ HELP cast through a raw pointer first
     b as usize; //~ ERROR non-scalar cast
     p as usize;
     //~^ ERROR casting
-    //~^^ HELP cast through a thin pointer
```
Comment (eddyb) [inline src/test/compile-fail/fat-ptr-cast.rs]: These should still emit the help. It's only when that `usize` is replaced with another type that it's wrong.

File: src/test/compile-fail/issue-21554.rs [hunk 3]
```diff
@@ -13,5 +13,4 @@ struct Inches(i32);
 fn main() {
     Inches as f32;
     //~^ ERROR casting
-    //~^^ cast through a usize first
```
Comment (eddyb) [inline src/test/compile-fail/issue-21554.rs]: How could this have disappeared?

**Comment by GuillaumeGomez** [comment]
@eddyb: Didn't pay attention to the tests I updated. Fixed now.

**Comment by eddyb** [comment]
@GuillaumeGomez Can you add non-`usize` cases to the existing tests, to show that they don't suggest?
IMO we should also never suggest anything for `NeedViaUsize`, who casts pointers to floating-point?!

**Comment by GuillaumeGomez** [comment]
That's another debate I think haha. If you can imagine something stupid, someone will do something stupider.

**Comment by GuillaumeGomez** [comment]
cc @eddyb

**Comment by arielb1** [comment]
`NeedViaUsize` was introduced because all directions of casts used to be OK under the pre-1.4.0 cast system. It can now be removed and replaced by `IllegalCast`.

**Pull request merged**